### PR TITLE
feat: event writer

### PR DESCRIPTION
**Pull request by johanandren**

### CODE REVIEW

**Review by patriknw — COMMENTED** [review]
looking good

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/EventWriter.scala [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+        val journal = Persistence(context.system).journalFor(journalPluginId)
+        context.log.debug("Event writer for journal [{}] starting up", journalPluginId)
+        apply(journal.toTyped)
+      })
+      .onFailure[Exception](SupervisorStrategy.restart)
+
+  def apply(journal: ActorRef[JournalProtocol.Message]): Behavior[Command] =
+    Behaviors
+      .setup[AnyRef] { context =>
+        // FIXME do we need to allow configuring per journal?
```
[the remark this answers is not in the document]
Comment (johanandren) [reply akka-persistence-typed/src/main/scala/akka/persistence/typed/EventWriter.scala:99]: Yeah, let's do that in the future if it turns out it is needed. 👍

**Comment by johanandren** [comment]
@patriknw Should I put it in `akka.persistence.typed.internal` rather than `akka.persistence.typed`?

**Comment by patriknw** [comment]
sounds good with `internal`, since that is what it will be

**Comment by johanandren** [comment]
Ready for final review

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventWriter.scala [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+      })
+      .onFailure[Exception](SupervisorStrategy.restart)
+
+  def apply(journal: ActorRef[JournalProtocol.Message]): Behavior[Command] =
+    Behaviors
+      .setup[AnyRef] { context =>
+        val maxBatchSize =
+          context.system.settings.config.getInt("akka.persistence.typed.event-writer.max-batch-size")
+        implicit val askTimeout: Timeout =
+          context.system.settings.config.getDuration("akka.persistence.typed.event-writer.ask-timeout").asScala
```
Comment (patriknw) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventWriter.scala:105]: maybe more idiomatic to have Settings that is passed in, and read once by the extension

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventWriter.scala [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+                          context.log.trace2(
+                            "Successfully wrote event persistence id [{}], sequence nr [{}]",
+                            pid,
+                            message.sequenceNr)
+                        waiting ! StatusReply.success(WriteAck(pid, sequenceNr))
+                        val newState = stateForPid.copy(waitingForReply = stateForPid.waitingForReply - sequenceNr)
+                        handleUpdatedStateForPid(pid, newState)
+                    }
+                  } else {
+                    if (context.log.isTraceEnabled) {
```
Comment (patriknw) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventWriter.scala:185]: this if is not needed

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventWriter.scala [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+                          case Success(unexpected) =>
+                            throw new IllegalArgumentException(
+                              s"Got unexpected reply from journal ${unexpected.getClass} for pid $pid")
+                          case Failure(exception) =>
+                            throw new RuntimeException(
+                              s"Error finding highest sequence number in journal for pid $pid",
+                              exception)
+                        }
+                      } else {
+                        if (context.log.isTraceEnabled)
```
Comment (patriknw) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventWriter.scala:236]: if not needed, shouldn't happen too often, and it's only saving an array allocation

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventWriter.scala [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+              Behaviors.same
+
+            case _ =>
+              // ignore all other journal protocol messages
+              Behaviors.same
+          }
+
+        Behaviors.receiveMessage {
+          case Write(persistenceId, sequenceNumber, event, metadata, tags, replyTo) =>
+            val payload = if (tags.isEmpty) event else Tagged(event, tags)
```
Comment (patriknw) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventWriter.scala:255]: 👍

**Review by patriknw — APPROVED** [review]
LGTM, with some minor suggestion

**Pull request merged**